# PhaseSeparation
This repository is written as a guidance for proteomics data processing. Use machine learning models to predict protein functions (phase separation) from amino acid sequences. 
## References
Lee J, Cho H, Kwon I. Phase separation of low-complexity domains in cellular function and disease. Exp Mol Med. 2022;54(9):1412-1422. 
\
Statistics of local complexity in amino acid sequences and sequence databases - ScienceDirect. Accessed June 6, 2024. 
\
Walsh I, Martin AJM, Di Domenico T, Tosatto SCE. ESpritz: accurate and fast prediction of protein disorder. Bioinformatics. 2012;28(4):503-509. 
\
CIDER: Resources to Analyze Sequence-Ensemble Relationships of Intrinsically Disordered Proteins - ScienceDirect. Accessed June 6, 2024. 
\
Lancaster AK, Nutter-Upham A, Lindquist S, King OD. PLAAC: a web and command-line application to identify proteins with prion-like amino acid composition. Bioinformatics. 2014;30(17):2501-2502. 
\
First-generation predictors of biological protein phase separation. Current Opinion in Structural Biology. 2019;58:88-96. 
\
A Concentration-Dependent Liquid Phase Separation Can Cause Toxicity upon Increased Protein Expression - ScienceDirect. Accessed June 6, 2024. 
\
Vernon RM, Chong PA, Tsang B, et al. Pi-Pi contacts are an overlooked protein feature relevant to phase separation. Shan Y, ed. eLife. 2018;7:e31486. 
\
Proteomic analysis of post-translational modifications | Nature Biotechnology. Accessed June 7, 2024. 
\
PhosphoSitePlus, 2014: mutations, PTMs and recalibrations | Nucleic Acids Research | Oxford Academic. Accessed June 7, 2024.
